nemo/collections/asr/parts/mixins/mixins.py

```diff
@@ -45,65 +45,153 @@ def _setup_tokenizer(self, tokenizer_cfg: DictConfig):
+                self.artifacts.pop('tokenizer.spe_tokenizer_vocab', None)
+                for akey in list(self.artifacts.keys()):
+                    if akey.startswith('tokenizer.tokenizers.'):
+                        self.artifacts.pop(akey)
+
+            tokenizers_dict = {}
+            for lang, tcfg in self.tokenizer_cfg['tokenizers'].items():
+                (tokenizer, model_path, vocab_path, spe_vocab_path,) = self._make_tokenizer(tcfg, lang)
+                tokenizers_dict[lang] = tokenizer
+                if hasattr(self, 'cfg'):
+                    with open_dict(self.cfg):
+                        self.cfg.tokenizer.tokenizers[lang]['dir'] = self.tokenizer_cfg['tokenizers'][lang]['dir']
+                        self.cfg.tokenizer.tokenizers[lang]['type'] = self.tokenizer_cfg['tokenizers'][lang]['type']
+
+                # pop these now that we don't need them
+                if "dir" in tcfg:
+                    self.tokenizer_cfg['tokenizers'][lang].pop('dir')
+                self.tokenizer_cfg['tokenizers'][lang].pop('type')
+                self.tokenizer_cfg['tokenizers'][lang].pop('hf_kwargs', {})
+
+            self.tokenizer = tokenizers.AggregateTokenizer(tokenizers_dict)
+        else:
+            logging.debug('_setup_tokenizer: detected a monolingual tokenizer')
+            # need to de-register any old artifacts
+            if hasattr(self, 'cfg'):
+                with open_dict(self.cfg):
+                    self.cfg.tokenizer.pop('tokenizers', None)
+            if hasattr(self, 'artifacts'):
+                for akey in list(self.artifacts.keys()):
+                    if akey.startswith('tokenizer.tokenizers.'):
+                        self.artifacts.pop(akey)
+            (self.tokenizer, self.model_path, self.vocab_path, self.spe_vocab_path,) = self._make_tokenizer(
+                self.tokenizer_cfg
+            )
+
+            # can safely pop now
+            if 'dir' in self.tokenizer_cfg:
+                self.tokenizer_cfg.pop('dir')
+            self.tokenizer_cfg.pop('type')
+            self.tokenizer_cfg.pop('hf_kwargs', {})
+
+    def _make_tokenizer(self, tokenizer_cfg: DictConfig, lang=None):
+
+        tokenizer_type = tokenizer_cfg.get('type')
+        tokenizer_dir = tokenizer_cfg.get('dir')
+
+        if tokenizer_type not in ['bpe', 'wpe']:
+            raise ValueError(
+                '`tokenizer.type` must be either `bpe` for SentencePiece tokenizer or' '`wpe` for BERT based tokenizer'
+            )
+
+        # defaults
+        model_path = None
+        vocab_path = None
+        spe_vocab_path = None
+
+        if tokenizer_type == 'bpe':
             # This is a BPE Tokenizer
-            if 'model_path' in self.tokenizer_cfg:
-                model_path = self.tokenizer_cfg.get('model_path')
+            if 'model_path' in tokenizer_cfg:
+                model_path = tokenizer_cfg.get('model_path')
+            else:
```

Comment:
I ... Don't like this. It's too convoluted. Let's do one thing - split this function into two - the old one for mono lang and a new one for multilang. Also there are various places with missing self.xyz = abc assignments. Those are necessary for various functionality , making it even more crucial to split this function and have explicit code paths for the two types of Tokenizers. You may ofc call the mono tokenizer setup function from your multilingual Tokenizers setup function if refactor works well.

Reply:
done